spec/safe_statements_spec.rb

```diff
@@ -1501,6 +1501,92 @@ def up
             end
           end
 
+          it "creates valid index when sub-partition exists with no child partitions" do
+            create_range_partitioned_table(:foos3, migration_klass)
+            create_range_partitioned_table(:foos3_sub, migration_klass)
+
+            ActiveRecord::Base.connection.execute(<<~SQL)
+              ALTER TABLE foos3
+              ATTACH PARTITION foos3_sub
+              FOR VALUES FROM ('2020-01-01') TO ('2020-02-01')
+            SQL
+
+            test_migration = Class.new(migration_klass) do
+              def up
+                safe_add_concurrent_partitioned_index :foos3, :updated_at
+              end
+            end
+
+            allow(ActiveRecord::Base.connection).to receive(:execute).and_call_original
+
+            aggregate_failures do
+              expect(ActiveRecord::Base.connection).to receive(:execute).with(/CREATE INDEX "index_foos3_on_updated_at" ON ONLY/).once.ordered
+              expect(ActiveRecord::Base.connection).to receive(:execute).with(/CREATE INDEX "index_foos3_sub_on_updated_at" ON ONLY/).once.ordered
+              expect(ActiveRecord::Base.connection).to receive(:execute).with(/ALTER INDEX .+\nATTACH PARTITION/).once.ordered
+              expect(ActiveRecord::Base.connection).to_not receive(:execute).with(/CREATE INDEX CONCURRENTLY/)
+            end
+
+            test_migration.suppress_messages { test_migration.migrate(:up) }
+
+            aggregate_failures do
+              %i[foos3 foos3_sub].each do |table|
+                indexes = ActiveRecord::Base.connection.indexes(table)
```

Comment:
Does this only include valid indexes?

Reply:
No it would include invalid indexes as well. I think I see what you're getting at, but the subject under test would raise an error if the parent index is invalid, so these indexes are guaranteed to be valid

Reply:
> the subject under test would raise an error if the parent index is invalid
Sure, but this is the test, and we can't assume the subject 😄